File: OutFreeEvent.cpp
//
// Created by Can Candan on 04/06/17.
//

#include "OutFreeEvent.h"
#include "OutInEvent.h"
#include "TaskOutArrivalEvent.h"
#include "easylogging++.h"

OutFreeEvent::OutFreeEvent(Simulation *pSimulation, Task *pTask, double time, OutputDevice *pDevice) : Event(pSimulation, pTask, time) {
    out=pDevice;
    name="OutFreeEvent";
}

void OutFreeEvent::process() {
    log();
    out->setIdle(true);
    if (!task->isFinished()) {
        sim->addToRRQueue(task);
    }

    Task* newTask= sim->popFromRRQueue();
    if (newTask) {
        newTask->setWaitOut(newTask->getWaitOut()+(time-newTask->getOutTime()));
        TaskOutArrivalEvent* event= new TaskOutArrivalEvent(sim, newTask, time);
        sim->schedule(event);
    }
}

File: OutFreeEvent.java
/**
 * Created by veraxmedax on 04/06/2017.
 */
public class OutFreeEvent extends Event {
    private OutputDevice outputDevice;

    public OutFreeEvent(Simulation sim, Task task, double time, OutputDevice outputDevice) {
        super(sim, task, time);
        this.outputDevice = outputDevice;
    }

    public OutputDevice getOutputDevice() {
        return outputDevice;
    }

    public void setOutputDevice(OutputDevice outputDevice) {
        this.outputDevice = outputDevice;
    }

    @Override
    void process() {
        getOutputDevice().setIdle(true);
        if (getTask().isFinished()) {
            Event taskFinishedEvent = new TaskFinishedEvent(getSim(), getTask(), getTime());
            getSim().schedule(taskFinishedEvent);
        } else {
            getSim().getRrQueue().add(getTask());
        }
        setTask(getSim().getRrQueue().peek());
        double newTime = getTask().getWaitOut() + getTime() - getTask().getOutTime();
        getTask().setWaitOut(newTime);
        Event taskOutArrivalEvent = new TaskOutArrivalEvent(getSim(), getTask(), getTime());
        getSim().schedule(taskOutArrivalEvent);
    }
}
